test/functional/feature_assumeutxo.py

```diff
@@ -159,20 +162,12 @@ def no_sync():
         self.connect_nodes(0, 1)
 
         self.log.info(f"Ensuring snapshot chain syncs to tip. ({FINAL_HEIGHT})")
-
-        def check_for_final_height():
-            chainstates = n1.getchainstates()
-            # The background validation may have completed before we run our first
-            # check, so accept a final blockheight from either chainstate type.
-            cs = chainstates.get('snapshot') or chainstates.get('normal')
-            return cs['blocks'] == FINAL_HEIGHT
-
-        wait_until_helper(check_for_final_height)
+        wait_until_helper(lambda: n1.getchainstates()['chainstates'][-1]['blocks'] == FINAL_HEIGHT)
         self.sync_blocks(nodes=(n0, n1))
 
         self.log.info("Ensuring background validation completes")
         # N.B.: the `snapshot` key disappears once the background validation is complete.
-        wait_until_helper(lambda: not n1.getchainstates().get('snapshot'))
+        wait_until_helper(lambda: len(n1.getchainstates()['chainstates']) != 1)
```

Comment:
I'm not sure I follow this substitution... once background validation completes, `chainman.GetAll()` should (at least eventually) return only one chainstate: the fully-validated snapshot chainstate. Maybe `s/!=/==`?

Reply:
Funny, I think as-written this succeeds *most of the time* because the check is racily performed before validation actually completes, which essentially makes the check exit early.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/28590#discussion_r1347388519

> Funny, I think as-written this succeeds _most of the time_ because the check is racily performed before validation actually completes, which essentially makes the check exit early.

Right is the opposite of the condition you'd expect. Probably I just wrote the check one way, then inverted when it failed assuming I'd made a mistake. But either of these checks are racy, so it makes more sense to be direct and check the new "validated" field. Updated in new push.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/28590#discussion_r1347392493

Actually thinking about it more, the change you suggested is not racy, and is a little better than just checking the "validated" field because it also ensures that the background chainstate disappears from the list after it is disabled. Also check you suggested is more consistent with the check done for the n2 node below on line 218